### PR TITLE
feat: add k8s metrics collector

### PR DESCRIPTION
**Pull request by ahmedElqutb**
Here we introduce a collector that collects k8s metrics from `custom.metrics.k8s.io/v1beta1/` 
and would save them in the bundle in the directory `metrics` as follows:
```
metrics
|_ <resource_type> # can be Namespace, Pod, Node etc.
   |_ <metric_name> # raw metric name truncated from the resource name as per custom.metrics.k8s.io/v1beta1/
      |_ <namespace>.json or <non_namespaced_object>.json # values for namespaced resources metrics are saved together in a file named after the namespace. For non-namespaced resources, each resource has their metric values in a separate file named after the resource
```

the collector executes multiple metric requests that have the format

```
type MetricRequest struct {
	// Namespace for which to collect the metric values, empty for non-namespaces resources.
	Namespace string `json:"namespace,omitempty" yaml:"namespace,omitempty"`
	// ObjectName for which to collect metric values, all resources when empty.
	// Note that for namespaced resources a Namespace has to be supplied regardless.
	ObjectName string `json:"objectName,omitempty" yaml:"objectName,omitempty"`
	// MetricName name of the MetricValueList as per the APIResourceList from
	// custom.metrics.k8s.io/v1beta1
	ResourceMetricName string `json:"resourceMetricName,omitempty" yaml:"resourceMetricName,omitempty"`
}
```
this might cause some duplication in case the same request, for the same object, or specific objects and a wildcard request for the object's kind present.

### CODE REVIEW

File: pkg/apis/troubleshoot/v1beta2/collector_shared.go [hunk 1]
```diff
@@ -27,6 +27,25 @@ type ClusterResources struct {
...
+// MetricsRequest the details of the MetricValuesList to be retrieved
+type MetricRequest struct {
+	// Namespace for which to collect the metric values, empty for non-namespaces resources.
+	Namespace string `json:"namespace,omitempty" yaml:"namespace,omitempty"`
+	// ObjectName for which to collect metric values, all resources when empty.
+	// Note that for namespaced resources a Namespace has to be supplied regardless.
+	ObjectName string `json:"objectName,omitempty" yaml:"objectName,omitempty"`
+	// MetricName name of the MetricValueList as per the APIResourceList from
+	// custom.metrics.k8s.io/v1beta1
+	ResourceMetricName string `json:"resourceMetricName,omitempty" yaml:"resourceMetricName,omitempty"`
```
Comment (mhrabovcin) [inline pkg/apis/troubleshoot/v1beta2/collector_shared.go]: This one is required, we can remove omit empty?

File: pkg/apis/troubleshoot/v1beta2/collector_shared.go [hunk 2]
```diff
@@ -275,6 +294,7 @@ type Collect struct {
 	ClusterInfo      *ClusterInfo      `json:"clusterInfo,omitempty" yaml:"clusterInfo,omitempty"`
 	ClusterResources *ClusterResources `json:"clusterResources,omitempty" yaml:"clusterResources,omitempty"`
 	Secret           *Secret           `json:"secret,omitempty" yaml:"secret,omitempty"`
+	MetricsRequest   *MetricsRequest   `json:"metricsRequest,omitempty" yaml:"metricsRequest,omitempty"`
```
Comment (mhrabovcin) [inline pkg/apis/troubleshoot/v1beta2/collector_shared.go]: `CustomMetrics`?

File: pkg/collect/collector.go [hunk 3]
```diff
@@ -120,6 +122,9 @@ func getCollectorName(c interface{}) string {
 		collector = "cluster-info"
 	case *CollectClusterResources:
 		collector = "cluster-resources"
+	case *CollectMetrics:
+		collector = "metrics"
```
Comment (mhrabovcin) [inline pkg/collect/collector.go]: `custom-metrics`

**Comment by ahmedElqutb** [comment]
a sample bundle can be obtained through:
https://github.com/mesosphere/dkp-diagnostics/tree/ashraf/add-k8s-metrics-collector-to-config, build locally and run against any cluster

File: pkg/collect/k8s_metrics.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		payload, err := json.MarshalIndent(list, "", "  ")
+		if err != nil {
+			return nil, errors.Wrap(err, "could not format readings")
+		}
+		output.SaveResult(c.BundlePath, relativePath, bytes.NewBuffer(payload))
+	}
+	return output, nil
+}
+
+func constructEndpoint(metricRequest troubleshootv1beta2.MetricRequest) (string, string, error) {
```
Comment (mhrabovcin) [inline pkg/collect/k8s_metrics.go:87]: Can we add unit test for this function with different inputs combinations (namespaces vs non-namespaces, with resource name)

File: pkg/collect/k8s_metrics.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	resultLists := make(map[string][]custom_metrics.MetricValue)
+	for _, metricRequest := range c.Collector.MetricRequests {
+		endpoint, metricName, err := constructEndpoint(metricRequest)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not construct endpoint")
+		}
+		response, err := c.Client.CoreV1().RESTClient().Get().AbsPath(endpoint).DoRaw(context.Background())
+		if err != nil {
+			return nil, errors.Wrap(err, fmt.Sprintf("could not query endpoint %s", endpoint))
+		}
```
Comment (mhrabovcin) [inline pkg/collect/k8s_metrics.go]: Could this be tested with fake client?

File: pkg/collect/collector.go (hunk 3 above)
Comment (ahmedElqutb) [reply pkg/collect/collector.go]: [c974cee](https://github.com/mesosphere/troubleshoot/pull/80/commits/c974ceeea41661505b1c8c8adca721f3384ccd66)

File: pkg/apis/troubleshoot/v1beta2/collector_shared.go (hunk 2 above)
Comment (ahmedElqutb) [reply pkg/apis/troubleshoot/v1beta2/collector_shared.go]: [c974cee](https://github.com/mesosphere/troubleshoot/pull/80/commits/c974ceeea41661505b1c8c8adca721f3384ccd66)

File: pkg/apis/troubleshoot/v1beta2/collector_shared.go (hunk 1 above)
Comment (ahmedElqutb) [reply pkg/apis/troubleshoot/v1beta2/collector_shared.go]: [c974cee](https://github.com/mesosphere/troubleshoot/pull/80/commits/c974ceeea41661505b1c8c8adca721f3384ccd66)

File: pkg/collect/k8s_metrics.go (hunk 4 above)
Comment (ahmedElqutb) [reply pkg/collect/k8s_metrics.go:87]: [423ab51](https://github.com/mesosphere/troubleshoot/pull/80/commits/423ab51a4757dad1c5891459575bf978c56fabe4)

File: pkg/collect/k8s_metrics.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+func (c *CollectMetrics) Collect(progressChan chan<- interface{}) (CollectorResult, error) {
+	output := NewResult()
+	resultLists := make(map[string][]custom_metrics.MetricValue)
+	for _, metricRequest := range c.Collector.MetricRequests {
+		endpoint, metricName, err := constructEndpoint(metricRequest)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not construct endpoint")
+		}
+		response, err := c.Client.CoreV1().RESTClient().Get().AbsPath(endpoint).DoRaw(context.Background())
```
Comment (mhrabovcin) [inline pkg/collect/k8s_metrics.go]: There is a context property provided to the collector config.

File: pkg/collect/k8s_metrics.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+		if err != nil {
+			return nil, errors.Wrap(err, fmt.Sprintf("could not query endpoint %s", endpoint))
```
Comment (mhrabovcin) [inline pkg/collect/k8s_metrics.go]:
```suggestion
			return nil, errors.Wrapf(err, "could not query endpoint %s", endpoint)
```

File: pkg/collect/k8s_metrics.go (hunk 6 above)
Comment (ahmedElqutb) [reply pkg/collect/k8s_metrics.go]: 423ab51

File: pkg/collect/k8s_metrics.go (hunk 7 above)
Comment (ahmedElqutb) [reply pkg/collect/k8s_metrics.go]: 423ab51

**Review by mhrabovcin — APPROVED** [review]

**Pull request merged**